src/main/java/world/bentobox/parkour/listeners/CourseRunnerListener.java

```diff
@@ -21,188 +19,274 @@
+
+    }
+
+    @EventHandler
+    public void onTeleport(PlayerTeleportEvent e) {
+        boolean shouldStopRun = switch (e.getCause()) {
+        case ENDER_PEARL, CHORUS_FRUIT, DISMOUNT, EXIT_BED -> false;
+        case COMMAND, PLUGIN, NETHER_PORTAL, END_PORTAL, SPECTATE, END_GATEWAY, UNKNOWN -> true;
+        };
+        if (shouldStopRun && parkourRunManager.timers().containsKey(e.getPlayer().getUniqueId())) {
+            User user = User.getInstance(e.getPlayer().getUniqueId());
+            if (parkourRunManager.checkpoints().containsKey(e.getPlayer().getUniqueId()) && user.isOnline()) {
+                user.notify("parkour.session-ended");
+            }
+            parkourRunManager.clear(e.getPlayer().getUniqueId());
+        }
+        // Check world - only apply flag actions to Parkour world and only if player is not actively running the course
+        if (e.getTo() == null // To can sometimes be null...
+                || !addon.inWorld(e.getTo())
+                || parkourRunManager.timers().containsKey(e.getPlayer().getUniqueId())) {
+            return;
+        }
+        // Handle flag action for players who are not running
+        Optional<Island> fromIsland = addon.getIslands().getIslandAt(e.getFrom());
+        Optional<Island> toIsland = addon.getIslands().getIslandAt(e.getTo());
+
+        if (fromIsland.isPresent() && toIsland.isPresent() && fromIsland.get().equals(toIsland.get())) {
+            // same island teleport
+            Island island = fromIsland.get();
+            User user = User.getInstance(e.getPlayer());
+            if (island.getFlag(addon.PARKOUR_CREATIVE) <= island.getRank(user)) {
+                user.setGameMode(GameMode.CREATIVE);
+            } else {
+                user.setGameMode(GameMode.SURVIVAL);
+            }
         }
+
     }
 
+
     /**
      * Prevent players from issuing commands during a run
+     *
      * @param e PlayerCommandPreprocessEvent
      */
     @EventHandler(priority = EventPriority.NORMAL)
     public void onVisitorCommand(PlayerCommandPreprocessEvent e) {
-        if (!timers.containsKey(e.getPlayer().getUniqueId()) || !e.getPlayer().getGameMode().equals(GameMode.SURVIVAL)
-                || e.getPlayer().isOp()) {
+        if (!parkourRunManager.timers().containsKey(e.getPlayer().getUniqueId())
+                || !e.getPlayer().getGameMode().equals(GameMode.SURVIVAL)
+                || e.getPlayer().hasPermission("parkour.mod.bypasscommandban")) {
             return;
         }
+
+        String command = e.getMessage();
+        for (String parkourAllowedCommand : addon.getSettings().getParkourAllowedCommands()) {
+            if (command.startsWith(parkourAllowedCommand)) {
+                return;
+            }
+        }
+        // Always allow using /<base command> quit
+        if (addon.getPlayerCommand().isPresent()) {
+            CompositeCommand cmd = addon.getPlayerCommand().get();
```

Comment:
to make sonar cloud shut up, refactor addon.getPlayerCommand() into a local variable

Reply:
It's a false positive. isPresent() is called before get(). I just make it as a false positive and sent it to Sonar for review.

Reply:
Yeah I know, it's probably a false positive as
GetPlayerCommand is called twice, and there is no guarantee that the same optional is returned